core/src/main/java/com/linecorp/armeria/server/RouteDecoratingService.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+import java.util.function.Function;
+
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+
+import io.netty.util.AttributeKey;
+
+/**
+ * Decorates a {@link Service} whose {@link Route} matches the {@link #route}.
+ *
+ * {@link RouteDecoratingService} is used for binding your a {@link Service} to multiple {@code decorator}s
+ * with {@link Route}s. e.g.
+ * <pre>{@code
+ * > Server server = new ServerBuilder()
+ * >     .service("/api/users",  userService)
+ * >     .decoratorUnder("/", loggingDecorator)
+ * >     .decoratorUnder("/api", authDecorator)
+ * >     .decoratorUnder("/api/users", traceDecorator)
+ * >     .build();
+ * }</pre>
+ *
+ * {@link VirtualHostBuilder} wraps each specified {@code decorator} with {@link RouteDecoratingService} and
+ * merge them into a {@link Router}. When a client requests to {@code /api/users/1},
+ * {@link InitialDispatcherService} will work the following steps for building decorators
+ * being matched to the given request:
+ *
+ * <ul>
+ *   <li>Finds all matched decorators from {@link InitialDispatcherService#router}
+ *       using {@link RoutingContext}</li>
+ *   <li>Put them into {@code pendingDecorators} with keeping the finding order</li>
+ *   <li>Finally, put the original {@link Service} at the end of {@code pendingDecorators}</li>
+ * </ul>
+ *
+ * <p>The request will go through the below decorators to reach the {@code userService}.
+ * <pre>{@code
+ *  request -> initialDispatcherService
+ *          -> loggingDecorator         -> routeDecoratingService
+ *          -> authDecorator            -> routeDecoratingService
+ *          -> traceDecorator           -> routeDecoratingService
+ *          -> userService
+ * }</pre>
+ */
+public final class RouteDecoratingService implements HttpService {
```

Comment:
I guess we need to hide this from the public API?

Reply:
Oops, sorry... I changed this `public` for just checking `javadoc` preview, I should've reverted that.